Sending more data after response is finished causes "thread panicked"

Hello,

The following case is tested on granian 2.5.5 (cpython 3.12.11, linux):
```python
async def app(scope, receive, send):
    assert scope['type'] == 'http'

    await send({
        'type': 'http.response.start',
        'status': 200,
        'headers': [
            (b'content-type', b'text/plain'),
        ]
    })

    # body
    await send({
        'type': 'http.response.body',
        'body': b'Hello, ',
        'more_body': False
    })
    await send({
        'type': 'http.response.body',
        'body': b'World!'
    })
```

The response is correct, further messages are ignored... but it will also shut down.

[INFO] Starting granian (main PID: 21519)
[INFO] Listening at: http://127.0.0.1:9000
[INFO] Spawning worker-1 with PID: 21520
[WARNING] ASGI Lifespan errored, continuing without Lifespan support (to avoid Lifespan completely use "asginl" interface)
[INFO] Started worker-1
[INFO] Started worker-1 runtime-1

<details>
<summary>thread '&lt;unnamed&gt;' panicked at src/asgi/io.rs:222:93:</summary>

> called `Option::unwrap()` on a `None` value
> stack backtrace:
>    0:     0x7f8e94565282 - <unknown>
>    1:     0x7f8e9449f213 - <unknown>
>    2:     0x7f8e94564ddf - <unknown>
>    3:     0x7f8e945650e3 - <unknown>
>    4:     0x7f8e94564a8d - <unknown>
>    5:     0x7f8e945904d5 - <unknown>
>    6:     0x7f8e94590469 - <unknown>
>    7:     0x7f8e9459111c - <unknown>
>    8:     0x7f8e93e57f3f - <unknown>
>    9:     0x7f8e93e5807b - <unknown>
>   10:     0x7f8e93e58228 - <unknown>
>   11:     0x7f8e94278f21 - <unknown>
>   12:     0x7f8e94d7bca0 - <unknown>
>   13:     0x7f8e94d7ac38 - PyObject_Vectorcall
>   14:     0x7f8e94c8b4d9 - <unknown>
>   15:     0x7f8e94d9a07c - <unknown>
>   16:     0x7f8e93ad4798 - <unknown>
>   17:     0x7f8e93ad54ea - <unknown>
>   18:     0x7f8e94d8d015 - _PyObject_Call
>   19:     0x7f8e92c34ce7 - <unknown>
>   20:     0x7f8e92c3d66c - <unknown>
>   21:     0x7f8e92c4a0bf - <unknown>
>   22:     0x7f8e94d854d0 - <unknown>
>   23:     0x7f8e94d7ac38 - PyObject_Vectorcall
>   24:     0x7f8e94c8b4d9 - <unknown>
>   25:     0x7f8e94da6fee - PyObject_VectorcallMethod
>   26:     0x7f8e93e6dfe3 - <unknown>
>   27:     0x7f8e94297b50 - <unknown>
>   28:     0x7f8e942e60b8 - <unknown>
>   29:     0x7f8e94d7bca0 - <unknown>
>   30:     0x7f8e94d7ac38 - PyObject_Vectorcall
>   31:     0x7f8e94c8b4d9 - <unknown>
>   32:     0x7f8e94d69adf - _PyObject_FastCallDictTstate
>   33:     0x7f8e94d898be - _PyObject_Call_Prepend
>   34:     0x7f8e94d896a0 - <unknown>
>   35:     0x7f8e94d678cb - <unknown>
>   36:     0x7f8e94d676ab - _PyObject_MakeTpCall
>   37:     0x7f8e94c8b4d9 - <unknown>
>   38:     0x7f8e94d9df2d - <unknown>
>   39:     0x7f8e94d8d14a - <unknown>
>   40:     0x7f8e94c8bb17 - <unknown>
>   41:     0x7f8e94d9df2d - <unknown>
>   42:     0x7f8e94d8d14a - <unknown>
>   43:     0x7f8e94c8bb17 - <unknown>
>   44:     0x7f8e94d69b70 - _PyObject_FastCallDictTstate
>   45:     0x7f8e94d898be - _PyObject_Call_Prepend
>   46:     0x7f8e94e31350 - <unknown>
>   47:     0x7f8e94d67744 - _PyObject_MakeTpCall
>   48:     0x7f8e94c8b4d9 - <unknown>
>   49:     0x7f8e94de82d3 - PyEval_EvalCode
>   50:     0x7f8e94e24557 - <unknown>
>   51:     0x7f8e94e1f78c - <unknown>
>   52:     0x7f8e94e1c2a3 - <unknown>
>   53:     0x7f8e94e1bda2 - _PyRun_SimpleFileObject
>   54:     0x7f8e94e1bae4 - _PyRun_AnyFileObject
>   55:     0x7f8e94e188a3 - Py_RunMain
>   56:     0x7f8e94dd1d57 - Py_BytesMain
>   57:     0x7f8e9529370a - <unknown>
> [ERROR] Unexpected exit from worker-1
> [INFO] Shutting down granian
> 

</details>

Thanks!

--- a/src/asgi/io.rs
+++ b/src/asgi/io.rs
@@ -218,31 +218,35 @@ impl ASGIHTTPProtocol {
                     more,
                     self.response_chunked.load(atomic::Ordering::Relaxed),
                 ) {
-                    (true, false, false) => {
-                        let (status, headers) = self.response_intent.lock().unwrap().take().unwrap();
-                        self.send_response(
-                            status,
-                            headers,
-                            http_body_util::Full::new(body::Bytes::from(body))
-                                .map_err(std::convert::Into::into)
-                                .boxed(),
-                        );
-                        self.flow_tx_waiter.notify_one();
-                        empty_future_into_py(py)
-                    }
-                    (true, true, false) => {
-                        self.response_chunked.store(true, atomic::Ordering::Relaxed);
-                        let (status, headers) = self.response_intent.lock().unwrap().take().unwrap();
-                        let (body_tx, body_rx) = mpsc::unbounded_channel::<body::Bytes>();
-                        let body_stream = http_body_util::StreamBody::new(
-                            tokio_stream::wrappers::UnboundedReceiverStream::new(body_rx)
-                                .map(body::Frame::data)
-                                .map(Result::Ok),
-                        );
-                        *self.body_tx.lock().unwrap() = Some(body_tx.clone());
-                        self.send_response(status, headers, BodyExt::boxed(body_stream));
-                        self.send_body(py, &body_tx, body, false)
-                    }
+                    (true, false, false) => match self.response_intent.lock().unwrap().take() {
+                        Some((status, headers)) => {
+                            self.send_response(
+                                status,
+                                headers,
+                                http_body_util::Full::new(body::Bytes::from(body))
+                                    .map_err(std::convert::Into::into)
+                                    .boxed(),
+                            );
+                            self.flow_tx_waiter.notify_one();
+                            empty_future_into_py(py)
+                        }
+                        _ => error_flow!("Response already finished"),
+                    },
+                    (true, true, false) => match self.response_intent.lock().unwrap().take() {
+                        Some((status, headers)) => {
+                            self.response_chunked.store(true, atomic::Ordering::Relaxed);
+                            let (body_tx, body_rx) = mpsc::unbounded_channel::<body::Bytes>();
+                            let body_stream = http_body_util::StreamBody::new(
+                                tokio_stream::wrappers::UnboundedReceiverStream::new(body_rx)
+                                    .map(body::Frame::data)
+                                    .map(Result::Ok),
+                            );
+                            *self.body_tx.lock().unwrap() = Some(body_tx.clone());
+                            self.send_response(status, headers, BodyExt::boxed(body_stream));
+                            self.send_body(py, &body_tx, body, false)
+                        }
+                        _ => error_flow!("Response already finished"),
+                    },
                     (true, true, true) => match &*self.body_tx.lock().unwrap() {
                         Some(tx) => self.send_body(py, tx, body, false),
                         _ => error_flow!("Transport not initialized or closed"),

--- a/tests/apps/asgi.py
+++ b/tests/apps/asgi.py
@@ -123,8 +123,26 @@ async def err_app(scope, receive, send):
     1 / 0
 
 
-async def err_proto(scope, receive, send):
-    await send({'type': 'wrong.msg'})
+async def err_proto_msg(scope, receive, send):
+    await send(PLAINTEXT_RESPONSE)
+    try:
+        await send({'type': 'wrong.msg'})
+    except Exception as e:
+        msg = e.args[0]
+    await send({'type': 'http.response.body', 'body': msg.encode('utf8'), 'more_body': False})
+
+
+async def err_proto_flow(scope, receive, send):
+    await send(PLAINTEXT_RESPONSE)
+    await send({'type': 'http.response.body', 'body': b'msg1', 'more_body': False})
+    try:
+        await send({'type': 'http.response.body', 'body': b'msg2', 'more_body': True})
+    except Exception:
+        pass
+    try:
+        await send({'type': 'http.response.body', 'body': b'msg3', 'more_body': False})
+    except Exception:
+        pass
 
 
 async def timeout_n(scope, receive, send):
@@ -172,7 +190,8 @@ def app(scope, receive, send):
         '/ws_echo': ws_echo,
         '/ws_push': ws_push,
         '/err_app': err_app,
-        '/err_proto': err_proto,
+        '/err_proto/type': err_proto_msg,
+        '/err_proto/flow': err_proto_flow,
         '/timeout_n': timeout_n,
         '/timeout_w': timeout_w,
     }.get(scope['path'], info)(scope, receive, send)

--- a/tests/test_asgi.py
+++ b/tests/test_asgi.py
@@ -75,9 +75,13 @@ async def test_app_error(asgi_server, runtime_mode):
 @pytest.mark.parametrize('runtime_mode', ['mt', 'st'])
 async def test_protocol_error(asgi_server, runtime_mode):
     async with asgi_server(runtime_mode, ws=False) as port:
-        res = httpx.get(f'http://localhost:{port}/err_proto')
+        res = httpx.get(f'http://localhost:{port}/err_proto/type')
+        assert res.status_code == 200
+        assert res.text == 'Unsupported ASGI message'
 
-    assert res.status_code == 500
+        res = httpx.get(f'http://localhost:{port}/err_proto/flow')
+        assert res.status_code == 200
+        assert res.text == 'msg1'
 
 
 @pytest.mark.asyncio

